dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/message/DefaultProgramMessageService.java

```diff
@@ -200,8 +201,10 @@ private boolean hasDataWriteAccess(ProgramMessage message) {
 
     if (message.hasEnrollment()) {
       object = message.getEnrollment().getProgram();
-    } else if (message.hasEvent()) {
-      object = message.getEvent().getProgramStage();
+    } else if (message.hasTrackerEvent()) {
+      object = message.getTrackerEvent().getProgramStage();
+    } else if (message.hasSingleEvent()) {
+      object = message.getSingleEvent().getProgramStage();
```

Comment:
Possibly something for future, we will not be having ProgramStage anymore in SingleEvents right?